doc/designs/quic-design/quic-record-layer.md

```diff
@@ -0,0 +1,353 @@
+ * immediately.
+ *
+ * If OSSL_QRL_TX_FLAG_DEFER_SEND is set, the packet is not sent immediately but
+ * is held until the next call to ossl_qrl_write_pkt. This can be used to
+ * concatenate several packets into a single transmitted datagram. Every packet
+ * but the final packet to be written into a datagram should have the flag
+ * unset. packet to be written into a datagram should have the flag unset.
+ */
+int ossl_qrl_write_pkt(OSSL_QRL *qrl, const OSSL_QRL_TX_PKT *pkt);
+
+/*
+ * Change the BIO being used by the QRL for TX.
+ */
+int ossl_qrl_set1_write_bio(OSSL_QRL *qrl, BIO *tx_bio);
+
+/* TODO MTU handling, etc. */
+```
+
+Packet Headers
+--------------
+
+Packet headers are logically represented by the following structure.
+
+This structure may be cleaned up a bit later and is being published here to
+provide some early guidance of what to expect.
+
+```c
+/*
+ * QUIC Packet Header
+ * ==================
+ *
+ * This structure provides a logical representation of a QUIC packet header.
+ *
+ * QUIC packet formats fall into the following categories:
+ *
+ *   Long Packets, which is subdivided into five possible packet types:
+ *     Version Negotiation (a special case);
+ *     Initial;
+ *     0-RTT;
+ *     Handshake; and
+ *     Retry
+ *
+ *   Short Packets, which comprises only a single packet type (1-RTT).
+ *
+ * The packet formats vary and common fields are found in some packets but
+ * not others. The below table indicates which fields are present in which
+ * kinds of packet. * indicates header protection is applied.
+ *
+ *   SLLLLL         Legend: 1=1-RTT, i=Initial, 0=0-RTT, h=Handshake
+ *   1i0hrv                 r=Retry, v=Version Negotiation
+ *   ------
+ *   1i0hrv         Header Form (0=Short, 1=Long)
+ *   1i0hr          Fixed Bit (always 1)
+ *   1              Spin Bit
+ *   1       *      Reserved Bits
+ *   1       *      Key Phase
+ *   1i0h    *      Packet Number Length
+ *    i0hr?         Long Packet Type
+ *    i0h           Type-Specific Bits
+ *    i0hr          Version (note: always 0 for Version Negotiation packets)
+ *   1i0hrv         Destination Connection ID
+ *    i0hrv         Source Connection ID
+ *   1i0h    *      Packet Number
+ *    i             Token
+ *    i0h           Length
+ *       r          Retry Token
+ *       r          Retry Integrity Tag
+ *
+ * For each field below, the conditions under which the field is valid are
+ * specified. If a field is not currently valid, its contents are undefined
+ * unless otherwise specified.
+ */
+#define QUIC_MAX_CONN_ID_LEN                20
+
+#define QUIC_V1_LONG_PKT_TYPE_INITIAL       0x0
+#define QUIC_V1_LONG_PKT_TYPE_0RTT          0x1
+#define QUIC_V1_LONG_PKT_TYPE_HANDSHAKE     0x2
+#define QUIC_V1_LONG_PKT_TYPE_RETRY         0x3
+
+typedef struct ossl_quic_pkt_hdr_st {
+    /* [ALL] 1 if this was a long packet, 0 otherwise. Always valid. */
+    unsigned int    is_long     :1;
```

Comment:
Does any part other than the record layer care?  I suspect that all other fields that are currently depending on this bit would simply have a value that fits both long and short packet headers:

-   the type field could have one of the values `initial`, `handshake`, `0rtt`, `1rtt` and `retry`, where `rtt1` is the value it gets automatically for short packets
-   the spin bit, key phase och other similar fields could simply be zero or false when they're not valid, or could be claimed valid only when the type is or isn't `1rtt`, as the case may be.

Reply:
Yeah, merging this and long_type is a good idea.

Reply:
With that change, `long_type` could probably just be called `type`